src/functions.rs

```diff
@@ -0,0 +1,348 @@
+/// [`Parameters`][] trait to consume those parameters and verify that you received the correct
+/// number and type of them.
+pub trait Function {
+    fn call(
+        &mut self,
+        graph: &mut Graph,
+        source: &str,
+        parameters: &mut dyn Parameters,
+    ) -> Result<Value, ExecutionError>;
+}
+
+/// A helper trait for consuming the parameters of a function.  You will typically use it as
+/// follows:
+///
+/// ```
+/// # use tree_sitter_graph::functions::Parameters;
+/// # use tree_sitter_graph::graph::Value;
+/// # use tree_sitter_graph::ExecutionError;
+/// # fn main() -> Result<(), ExecutionError> {
+/// # let param_vec = vec![Value::String("test".to_string()), Value::Integer(42)];
+/// # let mut params = param_vec.into_iter();
+/// let first_param = params.param()?.into_string()?;
+/// let second_param = params.param()?.into_integer()?;
+/// // etc
+/// params.finish()?;
+/// # Ok(())
+/// # }
+/// ```
+pub trait Parameters {
+    /// Returns the next parameter, returning an error if you have exhausted all of the parameters
+    /// that were passed in.
+    fn param(&mut self) -> Result<Value, ExecutionError>;
+
+    /// Ensures that there are no more parameters to consume.
+    fn finish(&mut self) -> Result<(), ExecutionError>;
+}
+
+impl<I> Parameters for I
+where
+    I: Iterator<Item = Value>,
+{
+    fn param(&mut self) -> Result<Value, ExecutionError> {
+        let value = self.next().ok_or(ExecutionError::InvalidParameters)?;
+        Ok(value)
+    }
+
+    fn finish(&mut self) -> Result<(), ExecutionError> {
+        if self.next().is_some() {
+            return Err(ExecutionError::InvalidParameters);
+        }
+        Ok(())
+    }
+}
+
+/// A library of named functions.
+#[derive(Default)]
+pub struct Functions {
+    functions: HashMap<Identifier, Box<dyn Function>>,
+}
+
+impl Functions {
+    /// Creates a new, empty library of functions.
+    pub fn new() -> Functions {
+        Functions::default()
+    }
+
+    /// Returns the standard library of functions, as defined in the [language
+    /// reference][`crate::reference::functions`].
+    pub fn stdlib(ctx: &mut Context) -> Functions {
+        let mut functions = Functions::new();
+        functions.add(ctx.add_identifier("child-index"), stdlib::ChildIndex);
+        functions.add(ctx.add_identifier("node"), stdlib::Node);
+        functions.add(ctx.add_identifier("plus"), stdlib::Plus);
+        functions.add(ctx.add_identifier("replace"), stdlib::Replace);
+        functions.add(ctx.add_identifier("source-text"), stdlib::SourceText);
+        functions.add(ctx.add_identifier("start-row"), stdlib::StartRow);
+        functions.add(ctx.add_identifier("start-column"), stdlib::StartColumn);
+        functions.add(ctx.add_identifier("end-row"), stdlib::EndRow);
+        functions.add(ctx.add_identifier("end-column"), stdlib::EndColumn);
+        functions.add(ctx.add_identifier("node-type"), stdlib::NodeType);
```

Comment:
@tausbn I brought in your changes from the `a-few-minor-fixes` branch.  I took the liberty of renaming the `kind` function to `node-type` — it's only called `kind` in the Rust bindings because `type` is a reserved word.  The underlying concept in tree-sitter itself is "[node type](https://tree-sitter.github.io/tree-sitter/using-parsers#syntax-nodes)".

Reply:
That's perfect, thank you! (Though actually I ended up not needing the `kind`/`node-type` function. I guess it's useful to keep around regardless, though.)